### PR TITLE
rebased PR44

### PR DESCRIPTION
**Pull request by gturri**
same as https://github.com/Zero-Waste-Paris/slack-agenda-app/pull/44 but without the commits already cherry-picked, and with subsequent conflicts resolved

### CODE REVIEW

File: DBAgenda.php [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            foreach($vCalendar->VEVENT->CATEGORIES as $category) {
+                $category = (string)$category;
+                
+                if(is_null($number_volunteers_required) and !is_null($number_volunteers_required = is_number_of_attendee_category($category))) {
+                    continue;
+                }
+            }
+        }
+        
+        $query = $this->pdo->prepare("REPLACE INTO events (vCalendarFilename, ETag, datetime_begin, number_volunteers_required, vCalendarRaw) VALUES (:vCalendarFilename, :ETag, :datetime_begin, :number_volunteers_required, :vCalendarRaw)");
```
Comment (gturri) [inline DBAgenda.php:196]: I wasn't aware of this `REPLACE` mysql statement. It looks very convenient! 👍

File: DBAgenda.php [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+            'datetime_begin' => $datetime_begin->format('Y-m-d H:i:s'),
+            'number_volunteers_required' => $number_volunteers_required,
+            'vCalendarRaw' => $event['vCalendarRaw']
+        ));
+        
+        if(isset($vCalendar->VEVENT->ATTENDEE)) {
+            foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                $mail = str_replace("mailto:", "", (string)$attendee);
+
+                $query = $this->pdo->prepare("SELECT * FROM attendees WHERE email=:email;");
```
Comment (gturri) [inline DBAgenda.php:209]: potential optimization for later: retrieve all the users we need in a single call, before the loop (but let's not worry about that for the moment)

**Review by gturri — COMMENTED** [review]
still work in progress, I haven't reviewed everything yet

File: DBAgenda.php [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                            'name' => $category
+                        ));
+                    } catch (PDOException $e) {
+                        if($e->errorInfo[0] === 23000 and $e->errorInfo[1] === 1062) { // Integrity constraint violation: 1062 Duplicate entry
+                            $this->log->warning($e->getMessage());
+                        } else {
+                            throw $e;
+                        }
+                    }
+                    $id = $this->getLastInsertedRowId();
```
Comment (gturri) [inline DBAgenda.php:269]: Will this work as expected in the case where we caught an exception? (I mean: in this case we failed to insert anything with this connection). Could it be safer to re-run `INSERT INTO categories (name) VALUES (:name);`?

**Review by gturri — COMMENTED** [review]
Cool, I've read it all!

First of all, thanks for those commits :)

That's a lot of comment but not everything is important. To put it in a nutshell, the items which seem the most important to me are:

- do we delete from the db the events deleted on the caldav server?
- I'm not sure (with the "throw / catch") that `getLastInsertedRowId` will always do what we want
- I'm pretty sure it would be safest to put the sql queries of `updateEvent` in a transaction
- the use of `localcache` in `removeDeletedEvents` seem weird to me

File: MySQLAgenda.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    vCalendarFilename               VARCHAR( 256 ),
+    email                           VARCHAR( 256 ),
+    FOREIGN KEY (email)             REFERENCES attendees(email),
+    FOREIGN KEY (vCalendarFilename) REFERENCES events(vCalendarFilename));");
+
+        $this->pdo->query("CREATE TABLE IF NOT EXISTS properties ( 
+    property                        VARCHAR( 256 ) PRIMARY KEY,
+    value                           VARCHAR( 256 ));");
+
+        $query = $this->pdo->prepare("INSERT IGNORE INTO properties (property, value) VALUES ('CTag', 'NULL')");
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply MySQLAgenda.php:64]: Yes indeed. However as we drop the database at first the `IGNORE` can be removed.

File: DBAgenda.php [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+
+        $results = $query->fetchAll(\PDO::FETCH_UNIQUE|\PDO::FETCH_ASSOC);
+        
+        foreach($results as $vCalendarFilename => &$result) {
+            $this->parseEvent($vCalendarFilename, $userid, $result);
+        }
+        return $results;
+    }
+    
+    public function parseEvent(string $vCalendarFilename, string $userid, array &$result) {
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: Indeed, 1735b5b

File: DBAgenda.php [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    protected function getCTag() {
+        $query = $this->pdo->prepare("SELECT * FROM properties WHERE property = :property");
+        $query->execute(array('property' => 'CTag'));
+        $result = $query->fetchAll();
+    }
+
+    protected function setCTag(string $CTag) {
+        $query = $this->pdo->prepare("UPDATE properties SET value=:value WHERE property=:property");
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php:117]: Yes, that's why this entry is inserted when creating the DB.

File: DBAgenda.php [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        $vCalendarFilename_to_update = [];
+        foreach($ETags as $vCalendarFilename => $remote_ETag) {
+            $vCalendarFilename = basename($vCalendarFilename);
+            
+            $query = $this->pdo->prepare("SELECT `vCalendarFilename`, `ETag` FROM events WHERE vCalendarFilename = :vCalendarFilename");
+            $query->execute(array('vCalendarFilename' => $vCalendarFilename));
+            $result = $query->fetchAll();
+            
+            if(count($result) === 0) {
+                $vCalendarFilename_to_update[] = $vCalendarFilename;
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php:135]: 8522197

File: DBAgenda.php [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+            } else {
+                $this->log->debug("no need to update $vCalendarFilename");
+            }
+        }
+        
+        if(count($vCalendarFilename_to_update) > 0) {
+            $this->updateEvents($vCalendarFilename_to_update);
+        }
+        
+        //$this->removeDeletedEvents($ETags);
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: I forgot to implement it ... 3640958

File: DBAgenda.php [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        foreach($ETags as $vCalendarFilename => $ETag) {
+            $vCalendarFilenames[] = basename($vCalendarFilename);
+        }
+        
+        foreach($this->localcache->getAllEventsNames() as $vCalendarFilename){
+            if(in_array($vCalendarFilename, $vCalendarFilenames)) {
+                $this->log->debug("No need to remove ". $vCalendarFilename);
+            } else {
+                $this->log->info("Need to remove ". $vCalendarFilename);
+                $this->localcache->deleteEvent($vCalendarFilename);
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: 3640958

File: DBAgenda.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: No, I forgot to implement it... be0bdb5

File: DBAgenda.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: be0bdb5

File: DBAgenda.php [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+        }
+    }
+    
+    protected function updateEvent(array $event) {
+        $vCalendar = \Sabre\VObject\Reader::read($event['vCalendarRaw']);
+        $datetime_begin = $vCalendar->VEVENT->DTSTART->getDateTime();
+
+        //$this->debug->log("deleting: $event[vCalendarFilename]");
+        
+        $query = $this->pdo->prepare("DELETE FROM `events_attendees` WHERE vCalendarFilename=:vCalendarFilename;");
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply DBAgenda.php]: 2b5c944 should fix that. 
Please double check, as I am not familiar with SQL transactions.

File: CalDAVClient.php [hunk 11]
```diff
@@ -167,7 +167,7 @@ function getETags() {
         // ...
         $data = [];
         foreach($service->parse($response) as $event) {
-            $data[$event['value']['href']] = trim($event['value']['propstat']['prop']['getetag'], '"');
+            $data[basename($event['value']['href'])] = trim($event['value']['propstat']['prop']['getetag'], '"');
```
[the remark this answers is not in the document]
Comment (vhouchouas) [reply CalDAVClient.php:170]: Indeed :) e2bfc10

**Pull request merged**